compiler/cli_reporter.nim

```diff
@@ -0,0 +1,3405 @@
+    if n.len > 1:
+      candidates.add("  first type mismatch at position: " & $err.arg)
+      # candidates.add "\n  reason: " & $err.firstMismatch.kind # for debugging
+      case err.kind:
+        of kUnknownNamedParam:
+          if nArg == nil:
+            candidates.add("\n  unknown named parameter")
+          else:
+            candidates.add("\n  unknown named parameter: " & $nArg[0])
+
+        of kAlreadyGiven:
+          candidates.add("\n  named param already provided: " & $nArg[0])
+
+        of kPositionalAlreadyGiven:
+          candidates.add("\n  positional param was already given as named param")
+
+        of kExtraArg:
+          candidates.add("\n  extra argument given")
+
+        of kMissingParam:
+          candidates.add("\n  missing parameter: " & nameParam)
+
+        of kTypeMismatch, kVarNeeded:
+          doAssert nArg != nil
+          let wanted = err.targetArg.typ
+          doAssert err.targetArg != nil
+
+          candidates.add("\n  required type for " & nameParam &  ": ")
+          candidates.addTypeDeclVerboseMaybe(conf, wanted)
+          candidates.add "\n  but expression '"
+
+          if err.kind == kVarNeeded:
+            candidates.add renderNotLValue(nArg)
+            candidates.add "' is immutable, not 'var'"
+
+          else:
+            candidates.add renderTree(nArg)
+            candidates.add "' is of type: "
+
+            let got = nArg.typ
+            candidates.addTypeDeclVerboseMaybe(conf, got)
+            doAssert wanted != nil
+
+            if got != nil:
+              if got.kind == tyProc and wanted.kind == tyProc:
+                # These are proc mismatches so,
+                # add the extra explict detail of the mismatch
+                candidates.addPragmaAndCallConvMismatch(wanted, got, conf)
+              effectProblem(wanted, got, candidates)
+
+        of kUnknown:
+          discard "do not break 'nim check'"
+
+      candidates.add "\n"
+      if err.arg == 1 and nArg.kind == nkTupleConstr and
+          n.kind == nkCommand:
+        maybeWrongSpace = true
+
+    candidatesAll.add candidates
+
+  candidatesAll.sort # fix #13538
+  candidates = join(candidatesAll)
+
+  if skipped > 0:
+    candidates.add(
+      $skipped &
+        " other mismatching symbols have been " &
+        "suppressed; compile with --showAllMismatches:on to see them\n")
+
+  if maybeWrongSpace:
+    candidates.add(
+      "maybe misplaced space between " & renderTree(n[0]) & " and '(' \n")
+
+  result = (prefer, candidates)
+
+proc presentSpellingCandidates*(
+  conf: ConfigRef, candidates: seq[SemSpellCandidate]): string =
+
+  result = "candidates (edit distance, scope distance); see '--spellSuggest':"
+  for candidate in candidates:
+    result.add "\n ($1, $2): '$3'" % [
+      $candidate.dist,
+      $candidate.depth,
+      $candidate.sym.name.s
+    ]
+
+    result.addDeclaredLoc(conf, candidate.sym)
+
+proc reportBody*(conf: ConfigRef, r: SemReport): string =
```

Comment:
I would put this proc as late as possible in the file because it's a beast, public API, and drowns out what follows.

Reply:
In fact I would use prototypes to shove the big ol' report body procs way at the bottom.